File: rust/src/leet_code/common/list_node.rs
// rust/src/leetcode/common/list_node.rs

use core::fmt;

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

impl fmt::Display for ListNode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // opening bracket
        write!(f, "list: [")?;

        let mut node = Some(self);
        let mut first = true;
        let mut count = 0;
        while let Some(n) = node {
            if !first {
                write!(f, ", ")?; // add comma between nodes
            }
            write!(f, "{}", n.val)?; // write current value
            node = n.next.as_deref();
            first = false;
            count += 1;
        }

        // closing bracket
        writeln!(f, "]")?;

        write!(f, "size: {}", count)?;

        Ok(())
    }
}

// Helper: convert Vec<i32> → Option<Box<ListNode>>
#[allow(dead_code)]
pub fn vec_to_list(nums: Vec<i32>) -> Option<Box<ListNode>> {
    let mut current = None;

    for &num in nums.iter().rev() {
        let mut node = Box::new(ListNode::new(num));
        node.next = current;
        current = Some(node);
    }

    current
}

// Helper: convert Option<Box<ListNode>> → Vec<i32>
#[allow(dead_code)]
pub fn list_to_vec(mut list: Option<Box<ListNode>>) -> Vec<i32> {
    let mut result = vec![];

    while let Some(node) = list {
        result.push(node.val);
        list = node.next;
    }

    result
}

File: cpp/src/leet_code/common/list_node.cpp
// cpp/src/leet_code/common/list_node.cpp

#include "list_node.hpp"

ListNode *vec_to_cyclic_list(const vector<int> &v, int pos)
{
    if (v.empty())
        return NULL;
    auto dummy = new ListNode(0);
    auto current = dummy;
    auto it = v.begin();
    for (; it != v.end(); it++)
    {
        current->next = new ListNode(*it);
        current = current->next;
    }
    auto head = dummy->next;
    delete dummy;
    if (pos >= 0)
    {
        auto node = head;
        for (int i = 0; i < pos; i++)
        {
            node = node->next;
        }
        current->next = node;
    }
    return head;
}